_episodes_rmd/10-functions.Rmd

```diff
@@ -165,10 +167,110 @@ kelvin_to_celsius <- function(temp) {
 > {: .solution}
 {: .challenge}
 
+## Interlude: Defensive Programming
 
-We're going to define
-a function that calculates the Gross Domestic Product of a nation from the data
-available in our dataset:
+Now that we've begun to appreciate how writing functions provides an efficient
+way to make R code re-usable and modular, we should note that it is important
+to ensure that functions only work in their intended use-cases. This  is a part
+of the concept of _defensive programming_. Defensive programming encourages us
+to frequently check conditions and throw an error if something is
+wrong. These checks are referred to as assertion statements because we want to
+assert some condition is `TRUE` before proceeding. This makes it easier to
+debug because we have a better idea of where the error originated.
+
+### Checking conditions with `stopifnot`
+
+Let’s start by re-examining `fahr_to_kelvin`, our function for converting
+temperatures from Fahrenheit to Kelvin. It was defined like so:
+
+```{r}
+fahr_to_kelvin <- function(temp) {
+  kelvin <- ((temp - 32) * (5 / 9)) + 273.15
+  return(kelvin)
+}
+```
+
+For this function to work as intended, the argument `temp` must be a `numeric`
+value; otherwise, the mathematical procedure for converting between the two
+temperature scales will not work. To create an error, we can use the function
+`stop`. For example, since the arugment `temp` must be a `numeric` vector, we
+could check for this condition with an `if` statement and throw an error if the
+condition was violated. We could augment our function above like so:
+
+```{r}
+fahr_to_kelvin <- function(temp) {
+  if (!is.numeric(temp)) {
+    stop("temp must be a numeric vector.")
+  }
+  kelvin <- ((temp - 32) * (5 / 9)) + 273.15
+  return(kelvin)
+}
+```
+
+If we had multiple conditions or arguments to check, it would take many lines
+of code to check all of them. Luckily R provides the convenience function
+`stopifnot`. We can list as many requirements that should evaluate to `TRUE`;
+`stopifnot` throws an error if it finds one that is `FALSE`. Listing these
+conditions also serves a secondary purpose as extra documentation for the
+function.
+
+Let's try out defensive programming with `stopifnot` by adding assertions to
+check the input to our function `fahr_to_kelvin`.
+
+We want to assert the following: `temp` is a numeric vector. We may do that like
+so:
+
+```{r}
+fahr_to_kelvin <- function(temp) {
+  stopifnot(is.numeric(temp))
+  kelvin <- ((temp - 32) * (5 / 9)) + 273.15
+  return(kelvin)
+}
+```
+
+It still works when given proper input.
+
+```{r}
+# freezing point of water
+fahr_to_kelvin(temp = 32)
+```
+
+But fails instantly if given improper input.
+
+```{r}
+# Metric is a factor instead of numeric
+fahr_to_kelvin(temp = as.factor(32))
+```
+
+> ## Challenge 3
+>
+> Use defensive programming to ensure that our `fahr_to_celcius` function throws
```

Comment:
I do believe it's an 's' and not a 'c' in Celsius, at least [according to wikipedia](https://en.wikipedia.org/wiki/Celsius). 

Also add `()`.